apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1342,21 +1377,265 @@ private void StartPush()
-                    Add(extendTask, Next(false, () => HandleAckResponse(extendTask, null, null, extends)));
+                    Add(extendTask, Next(false, _exactlyOnceDeliveryEnabled ? () => HandleAckResponseForExactlyOnce(extendTask, null, null, extends) : () => HandleAckResponse(extendTask, null, null, extends)));
                 }
                 if (nacks.Count > 0)
                 {
                     _pushInFlight += nacks.Count;
                     _concurrentPushCount += 1;
                     Task nackTask = _client.ModifyAckDeadlineAsync(_subscriptionName, nacks, 0, _hardStopCts.Token);
-                    Add(nackTask, Next(false, () => HandleAckResponse(nackTask, null, nacks, null)));
+                    Add(nackTask, Next(false, _exactlyOnceDeliveryEnabled ? () => HandleAckResponseForExactlyOnce(nackTask, null, nacks, null) : () => HandleAckResponse(nackTask, null, nacks, null)));
+                }
+            }
+
+            private void Retry(IEnumerable<string> temporaryFailures, Action<IEnumerable<string>> actionToRetry, int durationInSeconds = 600)
+            {
+                // TemporaryFailures may be empty but will never be null, so null check is not needed.
+                // By design, only items in temporaryFailures that exist in _retyrableIds will be retried.
+                // If id is not in _retryableIds dictionary, it is either already succeeded or not retryable.
+                // Failure may be non-retryable due to exceeded time duration or that failure may have become permanent.
+                if (temporaryFailures.Any())
+                {
+                    var retryableErrors = new ConcurrentBag<string>();
+                    // Not all temporaryFailures may need to be retried. See if the id exists in the dictionary and only then retry.
+                    foreach (var item in temporaryFailures.Where(j => _retryableIds.ContainsKey(j)))
+                    {
+                        retryableErrors.Add(item);
+                    }
+
+                    if (retryableErrors.Count == 0)
+                    {
+                        // Nothing to retry.
+                        return;
+                    }
+
+                    // All ids in retryableErrors will have the same backoff.
+                    // Get retryInfo for the first retryableError from the _retryableIds dictionary.
+                    if (!_retryableIds.TryGetValue(retryableErrors.First(), out var retryInfo))
+                    {
+                        // Nothing to retry.
+                        return;
+                    }
+
+                    var backoff = s_ackBackoff.NextBackoff(retryInfo.CurrentBackoff ?? TimeSpan.Zero);
+                    // If we have errors to retry and we are not past the specified duration, then retry.
+                    if (DateTime.UtcNow.Subtract(retryInfo.FirstTimeOfFailureInUtc).Add(backoff) <= TimeSpan.FromSeconds(durationInSeconds))
```

Comment:
I would change this to:

```csharp
if (DateTime.UtcNow <= retryInfo.FirstTimeOfFailureInUtc + TimeSpan.FromSeconds(durationInSeconds) - backoff)
```

... assuming that's what is meant. Possibly refactor in terms of a deadline?